src/par_iter/enumerate.rs

```diff
@@ -50,29 +51,50 @@ unsafe impl<M> ExactParallelIterator for Enumerate<M>
 impl<M> IndexedParallelIterator for Enumerate<M>
     where M: IndexedParallelIterator,
 {
-    type Producer = EnumerateProducer<M::Producer>;
+    fn with_producer<WP>(self, wp: WP) -> WP::Output
+        where WP: ProducerContinuation<Self::Item>
+    {
+        return self.base.with_producer(Continuation { wp: wp });
 
-    fn into_producer(self) -> (Self::Producer, <Self::Producer as Producer>::Shared) {
-        let (base, shared) = self.base.into_producer();
-        (EnumerateProducer { base: base, offset: 0 }, shared)
+        struct Continuation<WP> {
+            wp: WP,
+        }
+
+        impl<ITEM, WP> ProducerContinuation<ITEM> for Continuation<WP>
+            where WP: ProducerContinuation<(usize, ITEM)>
+        {
+            type Output = WP::Output;
+            fn with_producer<'produce, P>(self,
+                                          base: P,
+                                          shared: &'produce P::Shared)
+                                          -> WP::Output
+                where P: Producer<'produce, Item=ITEM>
+            {
+                let producer = EnumerateProducer { base: base,
+                                                   offset: 0,
+                                                   phantoms: PhantomType::new() };
+                self.wp.with_producer(producer, shared)
+            }
+        }
     }
 }
 
 ///////////////////////////////////////////////////////////////////////////
 // Producer implementation
 
-pub struct EnumerateProducer<M>
-    where M: Producer,
+pub struct EnumerateProducer<'p, P>
+    where P: Producer<'p>,
 {
-    base: M,
+    base: P,
     offset: usize,
+    phantoms: PhantomType<&'p ()>,
```

Comment:
Doesn't really matter here, but I used `PhantomType` because this does not represent data that is reachable which the compiler can't see.